net_ipc/tests/echo_ipc_test.rs

```diff
@@ -0,0 +1,200 @@
+                            }
+                            continue;
+                        }
+                        _ => panic!("not handled"),
+                    },
+                    None => panic!("timeout wating for data"),
+                },
+            }
+        };
+
+        // de-arc the cb_result variable
+        let cb_result = match Arc::try_unwrap(cb_result) {
+            Ok(c) => c.into_inner().unwrap().unwrap(),
+            _ => panic!("couldn't un-Arc"),
+        };
+
+        // if they are both errors - return a stringified error
+        if let Err(ref cb_e) = cb_result {
+            if let Err(ref msg_e) = msg {
+                let cb_e = format!("{}", cb_e);
+                let msg_e = format!("{}", msg_e);
+                if cb_e == msg_e {
+                    return Err(msg_e.lines().next().unwrap().to_string());
+                }
+            }
+        }
+
+        // if they are both successes - return a stringified success
+        if let Ok(ref cb_o) = cb_result {
+            if let Ok(ref msg_o) = msg {
+                let cb_o = String::from_utf8_lossy(&cb_o).to_string();
+                let msg_o = String::from_utf8_lossy(&msg_o).to_string();
+                if cb_o == msg_o {
+                    return Ok(msg_o);
+                }
+            }
+        }
+
+        // if they do not match, we need to panic!
+        panic!(
+            "callback and event did not match {:?} / {:?}",
+            cb_result, msg
+        );
+    }
+
+    /// cleanup both the nodejs echo-server and the ipc client connection
+    fn destroy(mut self) {
+        println!("attempting to kill echo server");
+        unsafe {
+            libc::kill(self.srv.id() as i32, libc::SIGTERM);
+        }
+        self.srv.wait().unwrap();
+        println!("echo server off");
+
+        println!("attempting to kill zeromq context");
+        self.cli.close().unwrap();
+        ZmqIpcClient::destroy_context().unwrap();
+        println!("zemomq is off");
+    }
+}
+
+#[test]
+fn it_can_send_call_and_call_resp() {
+    let mut frame = TestFrame::new();
+
+    println!("### BEGIN TEST SUITE ###");
+
+    println!("--- making `hello` call ---");
+    assert_eq!(Ok("echo: hello".to_string()), frame.execute(b"hello"));
+
+    println!("--- making `error` call ---");
+    assert_eq!(
+        Err("IpcError: Error: echo: error".to_string()),
+        frame.execute(b"error")
+    );
+
+    println!("--- making `call-hello` call ---");
+    assert_eq!(
+        Ok("srv-got: `echo: srv-hello`".to_string()),
+        frame.execute(b"call-hello")
+    );
+
+    println!("--- making `call-error` call ---");
+    assert_eq!(
+        Ok("srv-got: `Error: IpcError: echo: srv-error`".to_string()),
+        frame.execute(b"call-error")
+    );
+
+    println!("### TEST SUITE FINISHED - CLEANUP ###");
+
+    frame.destroy();
+
+    println!("### END TEST SUITE ###");
+}
```

Comment:
Serious integration testing, awesome!

Reply:
@lucksus 

Awesome! Thanks for the feedback : ) Just a couple responses:

> why do we need all of this instead of just opening a websocket

Using unix sockets we can get up to 7X the throughput of TCP (https://github.com/rigtorp/ipc-bench). And with ZMQ we don't loose the flexibility of being able to fall back to TCP if the OS doesn't support them.

> something that is easy to se-/deserialize, like strings and JSON

I figured we'd often be sending raw binary data to other peers on the network, so I went with binary / https://crates.io/crates/rmp... if we'd rather go with json, that's a super simple switch I could make in about 10 minutes, but it would add overhead of base64 if we are sending binary stuff.